File: C++/2-Medium/00130._被围绕的区域/00130._被围绕的区域.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    void solve(vector<vector<char>>& board) {
        if (board.empty()) return;
        int m = board.size(), n = board[0].size();
        
        // 边界开始的 DFS 搜索
        for (int i = 0; i < m; ++i) {
            dfs(board, i, 0);
            dfs(board, i, n - 1);
        }
        for (int j = 1; j < n - 1; ++j) {
            dfs(board, 0, j);
            dfs(board, m - 1, j);
        }
        
        // 转换和恢复
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == 'A') board[i][j] = 'O';
            }
        }
    }
    
    void dfs(vector<vector<char>>& board, int i, int j) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != 'O')
            return;
        board[i][j] = 'A'; // 标记为 A
        dfs(board, i - 1, j);
        dfs(board, i + 1, j);
        dfs(board, i, j - 1);
        dfs(board, i, j + 1);
    }
};

int main() {
    vector<vector<char>> board = {
        {'X', 'X', 'X', 'X'},
        {'X', 'O', 'O', 'X'},
        {'X', 'X', 'O', 'X'},
        {'X', 'O', 'X', 'X'}
    };
    Solution().solve(board);
    for (auto &row : board) {
        for (char c : row) cout << c << " ";
        cout << "\\n";
    }
    return 0;
}

File: Python/2-Medium/00130._被围绕的区域/00130._被围绕的区域.py
from typing import List

def solve(board: List[List[str]]) -> None:
    """
    采用 DFS 算法修改被 'X' 围绕的区域。
    
    Args:
    - board: List[List[str]], 二维字符数组，代表棋盘。
    
    Returns:
    None, 直接在输入的 board 上修改。
    """
    if not board or not board[0]:
        return

    m, n = len(board), len(board[0])

    def dfs(x: int, y: int) -> None:
        """
        深度优先搜索，标记与边界上 'O' 相连的 'O'。
        
        Args:
        - x: 当前行坐标。
        - y: 当前列坐标。
        """
        if x < 0 or x >= m or y < 0 or y >= n or board[x][y] != 'O':
            return
        board[x][y] = 'A'  # 标记当前位置
        dfs(x + 1, y)
        dfs(x - 1, y)
        dfs(x, y + 1)
        dfs(x, y - 1)

    # 从边界开始 DFS
    for i in range(m):
        dfs(i, 0)
        dfs(i, n - 1)
    for j in range(n):
        dfs(0, j)
        dfs(m - 1, j)

    # 将 'A' 复原为 'O'，未标记的 'O' 转换为 'X'
    for i in range(m):
        for j in range(n):
            if board[i][j] == 'A':
                board[i][j] = 'O'
            elif board[i][j] == 'O':
                board[i][j] = 'X'

# 示例棋盘
board_example = [
    ["X", "X", "X", "X"],
    ["X", "O", "O", "X"],
    ["X", "X", "O", "X"],
    ["X", "O", "X", "X"]
]
solve(board_example)
board_example